SConstruct

```diff
@@ -62,154 +81,98 @@ if GetOption("base_config") == "e2_xml":
+    # Launch and configure a new environment for each target BUILD_TARGETS
+    # contains plain commandline targets or the contents of -e2_target
+    # ONLY IF there are no commandline targets.
+    for target_raw in BUILD_TARGETS:
+        # Convert entry to plain string so there are no differences
+        target = str(target_raw)
+        # between plain commandline targets and "option" targets
+        multienv[target] = env.Clone(
+            BUILD_LABEL=target, BUILD_DIR=os.path.abspath(target)
         )
-    )
-    env["LINKFLAGS"] = cproject.get_link_flags(e2_target)
-    env["LINKCOM"] = "$CXX $CPPFLAGS -o $TARGET $SOURCES $LINKFLAGS"
-    
-    
-    env["LIBPATH"] = [str(i) for i in cproject.get_link_libs_order(e2_target)]
-    # log.debug(env["LIBPATH"])
+
+        if not multienv[target].SConscript(
+            os.path.join("site_scons", "e2_xml.scons"),
+            exports={
+                "env": multienv[target],
+                "log_parent": log,
+                "cproject": cproject,
+                "e2_build_targets": e2_build_targets,
+                "dbt_build_targets": dbt_build_targets,
+            },
+        ):
+            success = False
+            Return("success")
 
 # GENERIC ENVIRONMENT STUFF AND ACTIONS
 # In this area below the "if" block, we keep environment changes and actions that are
 # going to need doing regardless of the particular "mode" (eg: e2_xml).
 
-# Helper functions borrowed from FBT for wrapping lengthy commandlines into temp files.
-# This is primarily useful in windows, but shouldn't hurt anything on other platforms.
-wrap_tempfile(env, "LINKCOM")
-wrap_tempfile(env, "ARCOM")
-
-# Happy little custom builders and their actions.
-env.Append(
-    BUILDERS={
-        "HEXBuilder": Builder(
-            action=Action(
-                '${OBJCOPY} -O ihex "${SOURCE}" "${TARGET}"',
-                "${HEXCOMSTR}",
-            ),
-            suffix=".hex",
-            src_suffix=".elf",
-        ),
-        "BINBuilder": Builder(
-            action=Action(
-                '${OBJCOPY} -O binary -S "${SOURCE}" "${TARGET}"',
-                "${BINCOMSTR}",
-            ),
-            suffix=".bin",
-            src_suffix=".elf",
-        ),
-    }
-)
+# Iterate again, this time over configured environments. This should work regardless of
+# --base_config setting!
+for target, t_env in multienv.items():
+    # Helper functions borrowed from FBT for wrapping lengthy commandlines into temp files.
+    # This is primarily useful in windows, but shouldn't hurt anything on other platforms.
+    wrap_tempfile(t_env, "LINKCOM")
+    wrap_tempfile(t_env, "ARCOM")
+
+    # VariantDir does the magic to ensure output goes to the dbt- whatever
+    # build directory. Careful: THIS CAN BE REALLY FINICKY if paths aren't set right
+    VariantDir(
+        os.path.join(t_env["BUILD_LABEL"], REL_SOURCE_DIR), "#src", duplicate=False
```

Comment:
Is there a reason to use `BUILD_LABEL` instead of `BUILD_DIR` here? Given we're constructing a path the latter feels more appropriate (but it will generate a absolute path if that will cause the problems mentioned above?)

Reply:
VariantDir is, as noted VERY particular about what types of paths go where. I spent most of last weekend trying to get it to properly cooperate and this was the result. In this case, it prefers a path relative to the location of the SConstruct. It'd probably be a touch safer to os.path.relpath(t_env["BUILD_LABEL"]) so it tweaks the slashes as needed in case we ever have slashes (sub-subdirectories) in a label but that seems like a pretty edge case.

It's also very touchy if it gets run from within an SConscript file or as part of an SConscript() function call. 😿